src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +400,122 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
                         }
                     });
 
-                    var document = buffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
-                    var codeCleanupService = document.GetLanguageService<ICodeCleanupService>();
+                    var solution = await applyFixAsync(progressTracker, cancellationToken).ConfigureAwait(true);
 
-                    var allDiagnostics = codeCleanupService.GetAllDiagnostics();
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                    cancellationToken.ThrowIfCancellationRequested();
 
-                    var enabedDiagnosticSets = ArrayBuilder<DiagnosticSet>.GetInstance();
+                    return workspace.TryApplyChanges(solution, progressTracker);
+                }
+            }
+        }
 
-                    foreach (var diagnostic in allDiagnostics.Diagnostics)
-                    {
-                        foreach (var diagnosticId in diagnostic.DiagnosticIds)
-                        {
-                            if (context.EnabledFixIds.IsFixIdEnabled(diagnosticId))
-                            {
-                                enabedDiagnosticSets.Add(diagnostic);
-                                break;
-                            }
-                        }
-                    }
+        private async Task<Solution> FixSolutionAsync(Solution solution, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            // FixProjectAsync reports progress for documents within a project, so we use that information to update
+            // the solution progress.
+            var projectProgressTracker = new ProgressTracker((description, completedItems, totalItems) =>
+            {
+                progressTracker.Description = description;
 
-                    var isRemoveUnusedUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(RemoveUnusedImportsFixId);
-                    var isSortUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(SortImportsFixId);
-                    var enabledDiagnostics = new EnabledDiagnosticOptions(enabedDiagnosticSets.ToImmutableArray(),
-                        new OrganizeUsingsSet(isRemoveUnusedUsingsEnabled, isSortUsingsEnabled));
+                // The project progress tracker will report a TotalItems only up to the number of projects processed so
+                // far, but the total number of completed items will be correct. Assert these conditions in debug
+                // builds, but use bounds checks to ensure we don't throw exceptions at runtime.
+                Debug.Assert(completedItems <= progressTracker.TotalItems);
+                completedItems = Math.Min(progressTracker.CompletedItems, completedItems);
```

Comment:
is this sanity check stuff or there is case this min or max thing could happen? asking since I couldn't see case where it can happen.

Reply:
It is not expected to happen. The preceding comment notes that the assertions are present and not expected to fail for debug builds, and the bounds restriction is present to ensure we do not fail here at runtime.

Reply:
:arrow_right: This code has now been removed.